zbark/debark_test.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zbark
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/uber-go/zap"
+
+	"github.com/uber-common/bark"
+)
+
+func newDebark() (zap.Logger, *bytes.Buffer) {
+	buf := &bytes.Buffer{}
+	logger := logrus.New()
+	logger.Out = buf
+	logger.Level = logrus.DebugLevel
+	return Debarkify(bark.NewLoggerFromLogrus(logger), zap.DebugLevel), buf
+}
+
+func TestDebark_CastNoop(t *testing.T) {
+	orig := zap.NewJSON(zap.DebugLevel)
+	assert.True(t, orig == Debarkify(Barkify(orig), zap.DebugLevel))
+}
+
+var levels = []zap.Level{
+	zap.DebugLevel,
+	zap.InfoLevel,
+	zap.WarnLevel,
+	zap.ErrorLevel,
+}
+
+func TestDebark_Levels(t *testing.T) {
+	logger, _ := newDebark()
+	for _, l := range append(levels, zap.PanicLevel, zap.FatalLevel) {
+		logger.SetLevel(l)
+		assert.Equal(t, l, logger.Level())
+	}
+}
+
+func TestDebark_Check(t *testing.T) {
+	logger, _ := newDebark()
+	logger.SetLevel(zap.DebugLevel)
+	for _, l := range append(levels, zap.PanicLevel, zap.FatalLevel) {
+		assert.NotNil(t, logger.Check(l, "msg"))
+	}
+
+	logger.SetLevel(zap.FatalLevel)
+	for _, l := range append(levels, zap.PanicLevel) {
+		assert.Nil(t, logger.Check(l, "msg"))
+	}
+	assert.NotNil(t, logger.Check(zap.FatalLevel, "msg"))
+}
+
+func TestDebark_LeveledLogging(t *testing.T) {
+	logger, buf := newDebark()
+	logger.SetLevel(zap.DebugLevel)
+	for _, l := range levels {
+		buf.Reset()
+		assert.Equal(t, 0, buf.Len(), "buffer not zero")
+		logger.Log(l, "ohai")
+		assert.NotEqual(t, 0, buf.Len(), "%q did not log", l)
+		buf.Reset()
+	}
+
+	logger.SetLevel(zap.FatalLevel)
+	assert.Equal(t, 0, buf.Len(), "buffer not zero")
+	for _, l := range append(levels, zap.PanicLevel) {
+		logger.Log(l, "ohai")
```

Comment:
That's a good question. I think the answer is yes -- even if the logger is set to ignore panic logs, we should still panic, but skip logging the message.

I verified behaviours fatal/panic behaviours in a branch:
https://github.com/uber-go/zap/compare/pv_panic_always?expand=1

This actually found a minor issue with how the `Check` works for panics and fatals, but otherwise, the logger works as it should by default, it will skip the log but still panic/fatal.